xdp/src/umem.rs

```diff
@@ -0,0 +1,212 @@
+    }
+}
+
+pub struct SliceUmemFrame<'a> {
+    offset: usize,
+    len: usize,
+    _buf: PhantomData<&'a mut [u8]>,
+}
+
+impl SliceUmemFrame<'_> {
+    pub fn set_len(&mut self, len: usize) {
+        self.len = len;
+    }
+}
+
+impl Frame for SliceUmemFrame<'_> {
+    fn offset(&self) -> FrameOffset {
+        FrameOffset(self.offset)
+    }
+
+    fn len(&self) -> usize {
+        self.len
+    }
+}
+
+pub struct SliceUmem<'a> {
+    buffer: &'a mut [u8],
+    frame_size: u32,
+    available_frames: Vec<u64>,
+    capacity: usize,
+}
+
+impl<'a> SliceUmem<'a> {
+    pub fn new(buffer: &'a mut [u8], frame_size: u32) -> Result<Self, io::Error> {
+        debug_assert!(frame_size.is_power_of_two());
+        let capacity = buffer.len() / frame_size as usize;
+        Ok(Self {
+            available_frames: Vec::from_iter(0..capacity as u64),
+            capacity,
+            frame_size,
+            buffer,
+        })
+    }
+
+    pub fn capacity(&self) -> usize {
+        self.capacity
+    }
+
+    pub fn available(&self) -> usize {
+        self.available_frames.len()
+    }
+
+    pub fn reserve(&mut self) -> Option<SliceUmemFrame<'a>> {
+        let index = self.available_frames.pop()?;
+
+        Some(SliceUmemFrame {
+            offset: index as usize * self.frame_size as usize,
+            len: 0,
+            _buf: PhantomData,
+        })
+    }
+
+    pub fn release(&mut self, frame: FrameOffset) {
+        let index = frame.0 / self.frame_size as usize;
+        self.available_frames.push(index as u64);
+    }
+}
+
+impl<'a> Umem for SliceUmem<'a> {
+    type Frame = SliceUmemFrame<'a>;
+
+    fn as_ptr(&self) -> *const u8 {
+        self.buffer.as_ptr()
+    }
+
+    fn as_mut_ptr(&mut self) -> *mut u8 {
+        self.buffer.as_mut_ptr()
+    }
+
+    fn len(&self) -> usize {
+        self.buffer.len()
+    }
+
+    fn frame_size(&self) -> usize {
+        self.frame_size as usize
+    }
+}
+
+#[derive(Debug)]
+pub struct AllocError;
+
+pub struct PageAlignedMemory {
+    ptr: *mut u8,
+    len: usize,
+}
+
+impl PageAlignedMemory {
+    pub fn alloc(frame_size: usize, frame_count: usize) -> Result<Self, AllocError> {
+        Self::alloc_with_page_size(
+            frame_size,
+            frame_count,
+            // Safety: just a libc wrapper
+            unsafe { sysconf(_SC_PAGESIZE) as usize },
+            false,
+        )
+    }
+
+    pub fn alloc_with_page_size(
+        frame_size: usize,
+        frame_count: usize,
+        page_size: usize,
+        huge: bool,
+    ) -> Result<Self, AllocError> {
+        debug_assert!(frame_size.is_power_of_two());
+        debug_assert!(frame_count.is_power_of_two());
+        debug_assert!(page_size.is_power_of_two());
+        let memory_size = frame_count * frame_size;
+        let aligned_size = (memory_size + page_size - 1) & !(page_size - 1);
+
+        // Safety:
+        // doing an ANONYMOUS alloc. addr=NULL is ok, fd is not used.
+        let ptr = unsafe {
```

Comment:
We already pull in memmap2, maybe it has higher level API for this?

Reply:
it doesn't and stop trying to make me use more deps :P